File: lib/aura/kernel/git_state.rb
# frozen_string_literal: true

require "open3"
require "fileutils"

module Aura
  module Kernel
    class GitState
      def initialize(project_path)
        @project_path = project_path
      end

      def snapshot(tool_name, success: true)
        return unless git_repo?

        # We only snapshot on success by default
        return unless success

        message = "[Aura] Tool execution: #{tool_name}"

        # Add all changes (except state/ which should be gitignored)
        Open3.capture3("git", "add", ".", chdir: @project_path)

        # Check if there are changes to commit
        out, _err, = Open3.capture3("git", "status", "--porcelain", chdir: @project_path)
        return if out.to_s.strip.empty?

        # Commit
        Open3.capture3("git", "commit", "-m", message, chdir: @project_path)
      end

      private

      def git_repo?
        Dir.exist?(File.join(@project_path, ".git"))
      end
    end
  end
end

File: src/core/kernel/gitState.ts
import fs from 'fs';
import path from 'path';
import { execa } from 'execa';

export class GitState {
  private projectPath: string;

  constructor(projectPath: string) {
    this.projectPath = path.resolve(projectPath);
  }

  public async snapshot(toolName: string, success = true): Promise<void> {
    if (!this.isGitRepo()) {
      return;
    }

    if (!success) {
      return;
    }

    const message = `[Aura] Tool execution: ${toolName}`;

    try {
      // Add all changes (except state/ which should be gitignored)
      await execa('git', ['add', '.'], { cwd: this.projectPath });

      // Check status
      const { stdout } = await execa('git', ['status', '--porcelain'], { cwd: this.projectPath });
      if (!stdout.trim()) {
        return;
      }

      // Commit
      await execa('git', ['commit', '-m', message], { cwd: this.projectPath });
    } catch (e) {
      // Fail-safe
    }
  }

  private isGitRepo(): boolean {
    return fs.existsSync(path.join(this.projectPath, '.git'));
  }
}
